### PR TITLE
ncbi-toolkit: add version 22_0_0

### PR DESCRIPTION
**Pull request by omsai**

### CODE REVIEW

File: var/spack/repos/builtin/packages/ncbi-toolkit/package.py [hunk 1]
```diff
@@ -3,16 +3,23 @@
...
 
 
 class NcbiToolkit(AutotoolsPackage):
     """NCBI C++ Toolkit"""
 
     homepage = "https://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/"
-    url      = "ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/CURRENT/ncbi_cxx--21_0_0.tar.gz"
+    url      = "ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/CURRENT/ncbi_cxx--22_0_0.tar.gz"
+
+    version('22_0_0', 'e352d25b24c3a2d087c2cf3cedf6ea95')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/ncbi-toolkit/package.py]: It looks like this version can also be found in the ARCHIVE. Should we use that URL by default so the current one doesn't break when a new release comes out?

Comment (omsai) [reply var/spack/repos/builtin/packages/ncbi-toolkit/package.py]: Applied to 0a094fe

I confirmed that downloading 22_0_0 uses ARCHIVE and trying a new unreleased version automagically looks in CURRENT:

```console
$ spack fetch ncbi-toolkit@22_0_0
==> Fetching ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/ARCHIVE/2019/Mar_28_2019/ncbi_cxx--22_0_0.tar.gz
############################################################################################################################################ 100.0%
```

```console
$ spack fetch ncbi-toolkit@23_0_0
==> Warning: There is no checksum on file to fetch ncbi-toolkit@23_0_0 safely.
==>   Fetch anyway? [y/N] y
==> Fetching with no checksum.
  Add a checksum or use --no-checksum to skip this check.
==> Fetching ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/CURRENT/ncbi_cxx--23_0_0.tar.gz
#=O#-   #     #                                                                                                                                   
curl: (78) RETR response: 550
==> Fetching from ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/CURRENT/ncbi_cxx--23_0_0.tar.gz failed.
==> Error: All fetchers failed for ncbi-toolkit-23_0_0-p4pm3qovnyt6hrld2sm62rjws6hk73sw
```

**Comment by omsai** [comment]
1. While I have you, I was wondering if you had an opinion about the correctness of using underscores in the versioning scheme?  I feel it would be better to be able to `spack install ncbi-toolkit@22.0.0` instead of `22_0_0` because presumably having the underscores treats the version like a string which could complicate version comparisons?
2. I can create a separate pull request with `url_for_version()` or add it on to this one.
3. I don't know if I should somehow preserve compatibility with the underscore scheme?

**Comment by omsai** [comment]
What do you think of UConn-HPC/spack@d019c94?
1. It doesn't seem to break compatibility; it honors `spack install ncbi-toolkit@21.0.0` in addition to `spack install ncbi-toolkit@21_0_0` 
2. But it treats the fetches as different versions; instead of saying the package is already downloaded, it fetches a tarball for 21_0_0 and treats it as a separate version from 21.0.0.  That's probably a good thing?

**Comment by adamjstewart** [comment]
> 1. While I have you, I was wondering if you had an opinion about the correctness of using underscores in the versioning scheme?

Spack knows how to handle versions that use dots, dashes, or underscores, and treats them all equivalently, so there will be no problems with sorting or version comparison:
```python
$ spack python
Spack version 0.12.1
Python 3.7.3, Darwin x86_64
>>> from spack.version import *
>>> Version('1.2.3') == Version('1_2_3') == Version('1-2-3')
True
```
Since all ways of representing the version are identical, I usually default to whatever the tarball uses to distinguish the version number (1_2_3) unless it is advertised as 1.2.3 elsewhere on the homepage.

> 2. I can create a separate pull request with url_for_version() or add it on to this one.

Don't bother. Since every version contains the release date in the URL, there's no way to predict the URL based on the version. Just use a version-specific URL like it used to be.

**Comment by omsai** [comment]
> Don't bother. Since every version contains the release date in the URL, there's no way to predict the URL based on the version. Just use a version-specific URL like it used to be.

Sounds good!  Then I'm happy with the submission as is.

**Pull request merged**